### DISCUSSION TITLE
How to add a new library in webauthn-server-core?

### DISCUSSION THREAD

**Discussion by tarun14110** [Yubico/java-webauthn-server#282] [Q&A] [answered]
Hi, I am trying to extend functionality of `webauthn-server-core`. 
I have a local `jar` that I want to use as a library. I added `implementation(files("libtest/target/libtest.jar"))` in `webauthn-server-core/build.gradle.kts`. Then, I added import statement (`import org.testlib.*;`) in `RelyingParty.java`. 
When I try to build it using `./gradlew build` I get error 

> /Users/java-webauthn-server/webauthn-server-core/src/main/java/com/yubico/webauthn/RelyingParty.java:70: error: package org.testlib does not exist
> import org.testlib*;
> ^
> 1 error
> 
> > Task :webauthn-server-core:compileJava FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':webauthn-server-core:compileJava'.
> > Compilation failed; see the compiler error output for details.
> 
> * Try:
> > Run with --stacktrace option to get the stack trace.
> > Run with --info or --debug option to get more log output.
> > Run with --scan to get full insights.
> 
> * Get more help at https://help.gradle.org
> 
> Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
> 
> You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
> 
> See https://docs.gradle.org/8.1.1/userguide/command_line_interface.html#sec:command_line_warnings
> 
> BUILD FAILED in 6s
> 20 actionable tasks: 4 executed, 16 up-to-date

Am I following the right steps to add a dependency?

**Comment by emlun** [comment, marked as answer]
Hi! I think that should work, assuming your `libtest` project is a subdirectory of the `webauthn-server-core/` directory. Otherwise you'll have to use parent traversal (`../../libtest/target/libtest.jar`) or an absolute path to your `libtest` project directory.

You can run Gradle with the `--debug` option to check if your JAR makes it onto the classpath. For example:

```
$ ./gradlew :webauthn-server-core:cleanCompileJava
$ ./gradlew :webauthn-server-core:compileJava --debug
```

For me, this includes one extremely long line that starts with:

```
2023-05-16T11:10:45.177+0200 [DEBUG] [org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler] Compiler arguments: --release 8 [... rest of line omitted]
```

and contains a `-classpath` argument listing the JARs on the classpath during compilation. If your JAR is not listed there, then there's something wrong with the dependency configuration.
